### PR TITLE
Count distinct 1891

### LINKED ISSUE #1891: Wire up COUNT DISTINCT aggregate
From a syntax perspective it should look like this:

```sql
select count(distinct id) from events
```

The underlying implementation should use the DISTINCT Map function, but have a new reducer that combines the Map output and outputs a count.

### PR DESCRIPTION
**Pull request by corylanou**
- [x] Tests
- [x] Refactor several areas of new redundant code checks
- [x] Add test for `select count(distinct(too, many, arguments))`
- [x] Add test for `select count()`

Fixes #1891, https://github.com/influxdb/influxdb/issues/2596

### CODE REVIEW

File: influxql/ast.go [hunk 1]
```diff
@@ -968,6 +994,53 @@ func (s *SelectStatement) validateDistinct() error {
...
+				for _, a := range c.Args {
+					if _, ok := a.(*Distinct); ok {
+						return true
+					}
+				}
+			}
+		}
+	}
+	return false
+}
```
Comment (dgnorton) [inline influxql/ast.go]: Needs to check for `distinct(value)` also.  Line 1107 in the integration tests needs the FROM clause to catch this...
```
{s: `select count(distinct(too, many, arguments)) FROM myseries`, err: `count(distinct <field>) can only have one argument`},
```

Comment (corylanou) [inline influxql/ast.go]: Good catch.  Fixed.

File: influxql/ast.go [hunk 2]
```diff
@@ -914,6 +918,28 @@ func (s *SelectStatement) validate(tr targetRequirement) error {
 }
 
 func (s *SelectStatement) validateAggregates(tr targetRequirement) error {
+	// First, determine if specific calls have at least one and only one argument
+	for _, f := range s.Fields {
+		if c, ok := f.Expr.(*Call); ok {
+			switch c.Name {
+			case "derivative", "non_negative_derivative":
+				if exp, got := 1, len(c.Args); got < exp {
```
Comment (dgnorton) [inline influxql/ast.go]: should it be `got != exp`?

Comment (corylanou) [inline influxql/ast.go]: These can take 2 arguments, so it's checking that it got at least one.  Looking at it now it probably should read:

```go
if min, max, got := 1, 2, len(c.Args); got > max || got < min {
```

**Comment by dgnorton** [comment]
+1

**Pull request merged**